Address safer cpp warnings in Timer.h

cdumez opened the pull request:
#### a488ebde500e0506d470538e23bc3bd3bda20c0f
<pre>
Address safer cpp warnings in Timer.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=298848">https://bugs.webkit.org/show_bug.cgi?id=298848</a>

Reviewed by Ryosuke Niwa.

The Timer constructor used to capture a raw pointer to the object it will
call the &quot;timeout&quot; function on, which caused safer cpp warnings. To address
the issue, we now capture a WeakPtr / ThreadSafeWeakPtr / CheckedPtr as
appropriate, depending on what the object supports.

* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::weakRefCount const):
* Source/WTF/wtf/TypeTraits.h:
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeySession.h:
* Source/WebCore/Modules/geolocation/GeoNotifier.h:
* Source/WebCore/Modules/geolocation/GeolocationClient.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/accessibility/AXGeometryManager.h:
* Source/WebCore/bindings/js/GarbageCollectionController.h:
(WebCore::GarbageCollectionController::ref const): Deleted.
(WebCore::GarbageCollectionController::deref const): Deleted.
* Source/WebCore/dom/DocumentFontLoader.h:
* Source/WebCore/dom/DocumentParser.h:
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/parser/HTMLParserScheduler.h:
* Source/WebCore/html/track/VTTRegion.h:
* Source/WebCore/loader/NavigationScheduler.h:
* Source/WebCore/loader/cache/MemoryCache.h:
(WebCore::MemoryCache::TypeStatistic::TypeStatistic): Deleted.
(WebCore::MemoryCache::ref const): Deleted.
(WebCore::MemoryCache::deref const): Deleted.
(WebCore::MemoryCache::disabled const): Deleted.
(WebCore::MemoryCache::size const): Deleted.
(WebCore::MemoryCache::setDeadDecodedDataDeletionInterval): Deleted.
(WebCore::MemoryCache::deadDecodedDataDeletionInterval const): Deleted.
* Source/WebCore/page/ImageAnalysisQueue.h:
* Source/WebCore/page/PerformanceMonitor.h:
* Source/WebCore/page/SettingsBase.h:
* Source/WebCore/page/mac/ServicesOverlayController.h:
* Source/WebCore/page/scrolling/ScrollLatchingController.h:
* Source/WebCore/platform/Timer.h:
(WebCore::Timer::Timer):
* Source/WebCore/platform/gamepad/mac/HIDGamepadProvider.h:
(WebCore::HIDGamepadProvider::ref const): Deleted.
(WebCore::HIDGamepadProvider::deref const): Deleted.
(WebCore::HIDGamepadProvider::ignoreGameControllerFrameworkDevices): Deleted.
(WebCore::HIDGamepadProvider::numberOfConnectedGamepads const): Deleted.
* Source/WebCore/platform/gamepad/mac/HIDGamepadProvider.mm:
* Source/WebCore/platform/graphics/ImageFrameAnimator.h:
* Source/WebCore/platform/mediastream/RealtimeOutgoingVideoSource.h:
* Source/WebCore/platform/mock/TimerEventBasedMock.h:
* Source/WebCore/platform/network/DNSResolveQueue.cpp:
* Source/WebCore/platform/network/DNSResolveQueue.h:
* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.h:
* Source/WebCore/platform/network/curl/DNSResolveQueueCurl.cpp:
* Source/WebCore/platform/network/curl/DNSResolveQueueCurl.h:
* Source/WebCore/platform/network/glib/DNSResolveQueueGLib.cpp:
* Source/WebCore/platform/network/glib/DNSResolveQueueGLib.h:
* Source/WebCore/storage/StorageArea.h:
* Source/WebCore/style/MatchedDeclarationsCache.h:
* Source/WebCore/svg/animation/SMILTimeContainer.h:
* Source/WebCore/workers/WorkerAnimationController.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadMonitor.h:
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/Shared/WebMemorySampler.cpp:
* Source/WebKit/Shared/WebMemorySampler.h:
(WebKit::WebMemorySampler::ref const): Deleted.
(WebKit::WebMemorySampler::deref const): Deleted.
* Source/WebKit/webpushd/WebPushDaemon.h:
(WebPushD::WebPushDaemon::ref const): Deleted.
(WebPushD::WebPushDaemon::deref const): Deleted.
* Source/WebKit/webpushd/WebPushDaemon.mm:
* Source/WebKitLegacy/Storage/StorageAreaSync.h:
* Source/WebKitLegacy/WebCoreSupport/PingHandle.h:

Canonical link: <a href="https://commits.webkit.org/299985@main">https://commits.webkit.org/299985@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7b2481e1bc9bddd5608cbac9d7f352e8a20b485

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120893 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40587 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31243 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127303 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72969 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8b8deb74-9dc2-4a86-8f77-4ed8d0b49f90) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122769 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49164 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91819 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61069 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e3afa9d9-3af5-4268-b223-547bdff0b844) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123845 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32964 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108371 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72515 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/abea8db3-541b-4eff-9070-dd1af2c8da84) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31994 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26474 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70895 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/113019 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102460 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26653 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130161 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/119409 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47814 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36320 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100437 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48183 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104545 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100340 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25447 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45735 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23785 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44505 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47676 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53381 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/149547 "Build was cancelled. Recent messages:Failed to print configuration") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47147 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/149547 "Build was cancelled. Recent messages:Failed to print configuration") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50491 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48831 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/platform/Timer.h:
```diff
@@ -158,22 +158,30 @@ class Timer : public TimerBase {
-            (objectPtr->*function)();
+        : m_function([weakObject = ThreadSafeWeakPtr { object }, function] {
+            RefPtr protectedObject = weakObject.get();
+            (protectedObject.get()->*function)();
```
Shouldn't we be checking nullity of `protectedObject`?

cdumez replied:
We can, for robustness but the previous code would not have dealt with that either and at least it would crash safely now.

cdumez committed:
Address safer cpp warnings in Timer.h
https://bugs.webkit.org/show_bug.cgi?id=298848

Reviewed by Ryosuke Niwa.

The Timer constructor used to capture a raw pointer to the object it will
call the "timeout" function on, which caused safer cpp warnings. To address
the issue, we now capture a WeakPtr / ThreadSafeWeakPtr / CheckedPtr as
appropriate, depending on what the object supports.

* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::weakRefCount const):
* Source/WTF/wtf/TypeTraits.h:
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeySession.h:
* Source/WebCore/Modules/geolocation/GeoNotifier.h:
* Source/WebCore/Modules/geolocation/GeolocationClient.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/accessibility/AXGeometryManager.h:
* Source/WebCore/bindings/js/GarbageCollectionController.h:
(WebCore::GarbageCollectionController::ref const): Deleted.
(WebCore::GarbageCollectionController::deref const): Deleted.
* Source/WebCore/dom/DocumentFontLoader.h:
* Source/WebCore/dom/DocumentParser.h:
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/parser/HTMLParserScheduler.h:
* Source/WebCore/html/track/VTTRegion.h:
* Source/WebCore/loader/NavigationScheduler.h:
* Source/WebCore/loader/cache/MemoryCache.h:
(WebCore::MemoryCache::TypeStatistic::TypeStatistic): Deleted.
(WebCore::MemoryCache::ref const): Deleted.
(WebCore::MemoryCache::deref const): Deleted.
(WebCore::MemoryCache::disabled const): Deleted.
(WebCore::MemoryCache::size const): Deleted.
(WebCore::MemoryCache::setDeadDecodedDataDeletionInterval): Deleted.
(WebCore::MemoryCache::deadDecodedDataDeletionInterval const): Deleted.
* Source/WebCore/page/ImageAnalysisQueue.h:
* Source/WebCore/page/PerformanceMonitor.h:
* Source/WebCore/page/SettingsBase.h:
* Source/WebCore/page/mac/ServicesOverlayController.h:
* Source/WebCore/page/scrolling/ScrollLatchingController.h:
* Source/WebCore/platform/Timer.h:
(WebCore::Timer::Timer):
* Source/WebCore/platform/gamepad/mac/HIDGamepadProvider.h:
(WebCore::HIDGamepadProvider::ref const): Deleted.
(WebCore::HIDGamepadProvider::deref const): Deleted.
(WebCore::HIDGamepadProvider::ignoreGameControllerFrameworkDevices): Deleted.
(WebCore::HIDGamepadProvider::numberOfConnectedGamepads const): Deleted.
* Source/WebCore/platform/gamepad/mac/HIDGamepadProvider.mm:
* Source/WebCore/platform/graphics/ImageFrameAnimator.h:
* Source/WebCore/platform/mediastream/RealtimeOutgoingVideoSource.h:
* Source/WebCore/platform/mock/TimerEventBasedMock.h:
* Source/WebCore/platform/network/DNSResolveQueue.cpp:
* Source/WebCore/platform/network/DNSResolveQueue.h:
* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.h:
* Source/WebCore/platform/network/curl/DNSResolveQueueCurl.cpp:
* Source/WebCore/platform/network/curl/DNSResolveQueueCurl.h:
* Source/WebCore/platform/network/glib/DNSResolveQueueGLib.cpp:
* Source/WebCore/platform/network/glib/DNSResolveQueueGLib.h:
* Source/WebCore/storage/StorageArea.h:
* Source/WebCore/style/MatchedDeclarationsCache.h:
* Source/WebCore/svg/animation/SMILTimeContainer.h:
* Source/WebCore/workers/WorkerAnimationController.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadMonitor.h:
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/Shared/WebMemorySampler.cpp:
* Source/WebKit/Shared/WebMemorySampler.h:
(WebKit::WebMemorySampler::ref const): Deleted.
(WebKit::WebMemorySampler::deref const): Deleted.
* Source/WebKit/webpushd/WebPushDaemon.h:
(WebPushD::WebPushDaemon::ref const): Deleted.
(WebPushD::WebPushDaemon::deref const): Deleted.
* Source/WebKit/webpushd/WebPushDaemon.mm:
* Source/WebKitLegacy/Storage/StorageAreaSync.h:
* Source/WebKitLegacy/WebCoreSupport/PingHandle.h:

Canonical link: https://commits.webkit.org/299985@main
